### PR TITLE
Error when declaring a constant within another constant declaration

### LINKED ISSUE #2874: Compiler crash when defining a constant inside a constant
Code

This is how I accidentally discovered an issue. Not a very likely scenario right? It was a typo. I gave a variable a capitol letter which made it a const. :)

``` cr
FOO = begin
  BAR = 1
end

puts FOO
```

Here is a more simple example.

``` cr
A = B = 1

puts A
```

Output

```
Nil assertion failed
[4335949203] *raise<String>:NoReturn +163
[4336000352] *Nil#not_nil!<Nil>:NoReturn +16
[4345468781] *Crystal::ASTNode+@Crystal::ASTNode#transform<Crystal::ASTNode+, Crystal::CleanupTransformer>:Crystal::ASTNode+ +18381
[4345455796] *Crystal::ASTNode+@Crystal::ASTNode#transform<Crystal::ASTNode+, Crystal::CleanupTransformer>:Crystal::ASTNode+ +5396
[4345449632] *Crystal::CleanupTransformer#transform<Crystal::CleanupTransformer, Crystal::Expressions>:Crystal::Expressions +112
[4345451313] *Crystal::ASTNode+@Crystal::ASTNode#transform<Crystal::ASTNode+, Crystal::CleanupTransformer>:Crystal::ASTNode+ +913
[4345449632] *Crystal::CleanupTransformer#transform<Crystal::CleanupTransformer, Crystal::Expressions>:Crystal::Expressions +112
[4345448805] *Crystal::Program#cleanup<Crystal::Program, Crystal::Expressions>:Crystal::Expressions +341
[4337356536] *Crystal::Compiler#compile<Crystal::Compiler, Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +2552
[4346334439] *Crystal::Command#run_command<Crystal::Command, Bool>:Nil +151
[4336187718] *Crystal::Command#run<Crystal::Command>:(Array(Crystal::ImplementationTrace) | Array(Crystal::Init::View+:Class) | Array(String) | Bool | Crystal::Compiler::Result | Hash(String, String) | IO::FileDescriptor+ | Nil) +18774
[4336047297] main +17521

Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

Version

```
Crystal 0.18.2 (2016-06-17)
```

### PR DESCRIPTION
**Pull request by caspiano**
The following code will now raise a syntax error, as constants must be declared at the top level or within a type.
```crystal
A = B = 1
```

Resolves #2874

### CODE REVIEW

**Comment by HertzDevil** [comment]
Does this handle the snippet with the `begin`?

**Comment by caspiano** [comment]
@HertzDevil yep, this is caught by the dynamic constant declaration checking. I'll add a test case for it.

File: spec/compiler/semantic/const_spec.cr [hunk 1]
```diff
@@ -366,6 +366,15 @@ describe "Semantic: const" do
       "can't declare constant dynamically"
   end
 
+  it "errors on dynamic constant assignment inside a block assigned to a constant (#2874)" do
+    assert_error %(
+      A = do
```
Comment (Sija) [inline spec/compiler/semantic/const_spec.cr:371]:
```suggestion
      A = begin
```

Comment (caspiano) [reply spec/compiler/semantic/const_spec.cr:371]: Ah wow... I've been AFK for a bit. Cheers @Sija

**Comment by straight-shoota** [comment]
What happens if the constant assignment is nested in a call?
```cr
A = puts(B = 1)
```

**Review by straight-shoota — APPROVED** [review]

**Review by sdogruyol — APPROVED** [review]
Thank you @caspiano 🙏

**Pull request merged**